File: C/PRNG.c
#include <stdio.h>
#include <math.h>


#define ITER 1000
#define ESP 1e-9
//////////////////////////// SUPPORT FUNCTIONS ////////////////////////////
double sawtooth(double x, double epsilon){
	int l = (int)floor((x / epsilon + 1.0) / 2.0);
	if(x > (2*l+1)*epsilon || x < (2*l-1)*epsilon) return NAN;

	return pow(-1.0, l) * (x - 2 * l * epsilon);
}

double find_max(double vector[3]) {
    double max = fabs(vector[0]);
    for (int i = 1; i < 3; i++) {
        if (fabs(vector[i]) > max)
            max = fabs(vector[i]);
    }
    return max;
}

// Trả về giá trị eigenvalue lớn nhất
double spectralRadius(double A[3][3]) {
    double init_vector[3] = {1, 1, 1};
    double out_vector[3] = {0, 0, 0};

    double prev = 0;
    for (int i = 0; i < ITER; i++) {
        out_vector[0] = A[0][0] * init_vector[0] + A[0][1] * init_vector[1] + A[0][2] * init_vector[2];
        out_vector[1] = A[1][0] * init_vector[0] + A[1][1] * init_vector[1] + A[1][2] * init_vector[2];
        out_vector[2] = A[2][0] * init_vector[0] + A[2][1] * init_vector[1] + A[2][2] * init_vector[2];

        double max = find_max(out_vector);

        init_vector[0] = out_vector[0] / max;
        init_vector[1] = out_vector[1] / max;
        init_vector[2] = out_vector[2] / max;

        if (fabs(max - prev) < ESP) {
            return max;  // trả luôn
        }
        prev = max;
    }
    return prev;  // nếu hết vòng vẫn trả
}

//////////////////////////// SUPPORT FUNCTIONS ////////////////////////////


// Chaotic PRNG
// A: 3x3 matrix (Stable Schur matrix)
// sample: number of samples to generate
// N: burn-in iterations (skip transient)
// eps: parameter of sawtooth function
// x0: initial state (3 values)
// seq_out: pointer to array [sample][3]
void chaotic_PRNG(double A[3][3], int sample, int N, double epsilon, double x0[3], double (*seq)[3]){
	//Eigenvalues & spectral radius
	double spectral_radius = spectralRadius(A);

	if(spectral_radius >= 1) {
		printf("Matrix A is not stable\n");
		return;
	}

	// compute inf norm 
	double norm_inf_manual = 0.0;
    	for (int i = 0; i < 3; i++) {
		double row_sum = 0.0;

		for (int j = 0; j < 3; j++) 	row_sum += fabs(A[i][j]);

		if (row_sum > norm_inf_manual) 	norm_inf_manual = row_sum;
	}

	// sigma + err
	double err = 0.1;
	double sigma = fmax(3 * norm_inf_manual, sqrt(norm_inf_manual + 1.0)+1.0) + err;
	//printf("%.10f\n", sigma);

	//init
	double x[3];
    	for (int i = 0; i < 3; i++) x[i] = x0[i];	

	int k = 0;
	for(int step = 0; step < N + sample; step++){
		double U[3];
		for(int i = 0; i < 3; i++){
			U[i] = sawtooth(sigma * x[i], epsilon);
		}

		// x_next = A*x + U
		double x_next[3] = {0,0,0};
		for(int i = 0; i < 3; i++){
			for(int j = 0; j < 3; j++){
				x_next[i] += A[i][j] * x[j];
			}
			x_next[i] += U[i];
		}

		// update
		for(int i = 0; i < 3; i++) x[i] = x_next[i];
		
		if(step >= N){  // N initial iterations
			for(int i = 0; i < 3; i++) seq[k][i] = x[i];
			k++;
		}
	}
}


// ------- Main -------
/*
int main() {
    double A[3][3] = {
        {0.0, 1.0/2, 1.0/20},
        {1.0/3, 0.0, 1.0/3},
        {1.0/20, 1.0/2, 0.0}
    };

    int sample = 2000;
    int N = 100;

    double seq1[sample][3];
    double seq2[sample][3];

    double x0_1[3] = {0.1, 0.01, 0.0};
    double x0_2[3] = {0.10000001, 0.01, 0.0};

    chaotic_PRNG(A, sample, N, 0.05, x0_1, seq1);
    chaotic_PRNG(A, sample, N, 0.05, x0_2, seq2);

    // Xuất file 1
    FILE *f1 = fopen("./../outputsFromC/output1.csv", "w");
    for (int i = 0; i < sample; i++) {
        fprintf(f1, "%f,%f,%f\n", seq1[i][0], seq1[i][1], seq1[i][2]);
    }
    fclose(f1);

    // Xuất file 2
    FILE *f2 = fopen("./../outputsFromC/output2.csv", "w");
    for (int i = 0; i < sample; i++) {
        fprintf(f2, "%f,%f,%f\n", seq2[i][0], seq2[i][1], seq2[i][2]);
    }
    fclose(f2);

    printf("----------------- DONE -------------------\n");
    return 0;
}
*/

File: PRNG.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D


# sawtooth function
def phi_eps(x, eps=0.05):
    l = np.floor((x / eps + 1) / 2).astype(int)
    return ((-1.0) ** l) * (x - 2.0 * l * eps)

# Pseudo Random Number Generator
# A:        Stable Schur matrix
# sample:   The number of sample we need for generating S-box
# x0:       Initial vector of state variables
# N :       The first N iterations to avoid the transient state
def chaotic_PRNG(A, sample=2000, N = 100, eps=0.05, x0=(0.1, 0.01, 0.0)):
    # Eigenvalues & spectral radius
    eigenvalues = np.linalg.eigvals(A)
    spectral_radius = max(abs(eigenvalues))
    
    # inf norm
    row_sums = np.sum(np.abs(A), axis=1)
    norm_inf_manual = np.max(row_sums)

    # sigma + err
    err = 0.1
    sigma = max(3*norm_inf_manual, np.sqrt(norm_inf_manual+1.0)+1.0) + err

    # init 
    x = np.array(x0, dtype=float)
    seq = []

    for i in range(N + sample):
        U = np.array([
            phi_eps(sigma * x[0], eps),
            phi_eps(sigma * x[1], eps),
            phi_eps(sigma * x[2], eps)
        ])
        x_next = A @ x + U
        x = x_next

        if i >= N:
            seq.append(x.copy())

    return np.array(seq)

# -------------------------------
if __name__ == "__main__":
    A = np.array([[0.0, 1/2, 1/20],
                  [1/3, 0.0, 1/3],
                  [1/20, 1/2, 0.0]], dtype=float)

    seq = chaotic_PRNG(A)

# --------- Figure 3: Allure of the state variables of the proposed chaotic system ---------
    fig, axes = plt.subplots(3, 1, figsize=(10, 6), sharex=True)
    axes[0].plot(seq[:,0])
    axes[0].set_ylabel("x1(t)")

    axes[1].plot(seq[:,1])
    axes[1].set_ylabel("x2(t)")

    axes[2].plot(seq[:,2])
    axes[2].set_ylabel("x3(t)")
    axes[2].set_xlabel("Samples")



# --------- Figure 3: Chaotic trajectories ---------
    fig = plt.figure(figsize=(8,6))
    ax = fig.add_subplot(111, projection='3d')
    ax.plot(seq[:,0], seq[:,1], seq[:,2], lw=0.5)
    ax.set_xlabel("x1")
    ax.set_ylabel("x2")
    ax.set_zlabel("x3")
    ax.set_title("Chaotic Trajectories (Figure 3 style)")
    plt.show()

# --------- Figure 4: Sensitivity to initial conditions ---------
    seq1 = chaotic_PRNG(A, sample=100, eps=0.05, x0=(0.1, 0.01, 0.0))
    seq2 = chaotic_PRNG(A, sample=100, eps=0.05, x0=(0.10000001, 0.01, 0.0))

    plt.figure(figsize=(10,5))
    plt.plot(seq1[:,0], label="x0=(0.1,0.01,0.0)", lw=1)
    plt.plot(seq2[:,0], label="x0'=(0.10000001,0.01,0.0)", lw=1, linestyle="--")
    plt.xlabel("Samples")
    plt.ylabel("x1(t)")
    plt.title("Sensitivity to Initial Conditions (Figure 4 style)")
    plt.legend()
    plt.show()
